src/services/obs/dir_stream.rs

```diff
@@ -0,0 +1,259 @@
+
+                    if object.key.ends_with('/') {
+                        continue;
+                    }
+
+                    let de = DirEntry::new(
+                        backend.clone(),
+                        ObjectMode::FILE,
+                        &backend.get_rel_path(&object.key),
+                    );
+
+                    debug!(
+                        "dir object {} got entry, mode: {}, path: {}",
+                        &self.path,
+                        de.mode(),
+                        de.path(),
+                    );
+                    return Poll::Ready(Some(Ok(de)));
+                }
+
+                if self.done {
+                    debug!("dir object {} list done", &self.path);
+                    return Poll::Ready(None);
+                }
+
+                self.state = State::Idle;
+                self.poll_next(cx)
+            }
+        }
+    }
+}
+
+#[derive(Default, Debug, Deserialize)]
+#[serde(default, rename_all = "PascalCase")]
+struct Output {
+    name: String,
+    prefix: String,
+    contents: Vec<Content>,
+    common_prefixes: Option<Vec<CommonPrefix>>,
+    marker: String,
+    next_marker: Option<String>,
+}
+
+#[derive(Default, Debug, Deserialize)]
+#[serde(default, rename_all = "PascalCase")]
+struct CommonPrefix {
+    prefix: String,
+}
+
+#[derive(Default, Debug, Deserialize)]
+#[serde(default, rename_all = "PascalCase")]
+struct Content {
+    key: String,
+    size: u64,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bytes::Buf;
+
+    #[test]
+    fn test_parse_xml() {
+        let bs = bytes::Bytes::from(
+            r#"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<ListBucketResult xmlns="http://obs.cn-north-4.myhuaweicloud.com/doc/2015-06-30/">
+    <Name>examplebucket</Name>
+    <Prefix>obj</Prefix>
+    <Marker>obj002</Marker>
+    <NextMarker>obj004</NextMarker>
+    <MaxKeys>1000</MaxKeys>
+    <IsTruncated>false</IsTruncated>
+    <Contents>
+        <Key>obj002</Key>
+        <LastModified>2015-07-01T02:11:19.775Z</LastModified>
+        <ETag>"a72e382246ac83e86bd203389849e71d"</ETag>
+        <Size>9</Size>
+        <Owner>
+            <ID>b4bf1b36d9ca43d984fbcb9491b6fce9</ID>
+        </Owner>
+        <StorageClass>STANDARD</StorageClass>
+    </Contents>
+    <Contents>
+        <Key>obj003</Key>
+        <LastModified>2015-07-01T02:11:19.775Z</LastModified>
+        <ETag>"a72e382246ac83e86bd203389849e71d"</ETag>
+        <Size>10</Size>
+        <Owner>
+            <ID>b4bf1b36d9ca43d984fbcb9491b6fce9</ID>
+        </Owner>
+        <StorageClass>STANDARD</StorageClass>
+    </Contents>
+    <CommonPrefixes>
+        <Prefix>hello</Prefix>
+    </CommonPrefixes>
+    <CommonPrefixes>
+        <Prefix>world</Prefix>
+    </CommonPrefixes>
+</ListBucketResult>"#,
+        );
+        let out: Output = de::from_reader(bs.reader()).expect("must success");
+        println!("{:?}", out);
```

Comment:
Remove this line as we don't want `println` in testing.

Reply:
removed